### PR TITLE
Print initial validation loss + final validation loss

### LINKED ISSUE #1221: Calculate loss at beginning and end of training
I talked to some users who were a bit confused / surprised that the validation loss didn't get calculated during a run. I think that's a valid issue if you have a small dataset where your training steps are smaller than the default `eval.interval=100`.

I think the best way to address this is to also always calculate the validation loss always at the beginning and end of each training run. These would also be useful reference points to have in general. 

Should be easy for me to add.

### PR DESCRIPTION
**Pull request by rasbt**
EDIT: After Adrian already added the final validation, I greatly simplified this PR. So this only adds the initial validation. Since there was not that much enthusiasm from others, I made it optional and disabled it by default. I think it's a super useful feature to be able to know what the initial val_loss of the model is so that one can compare it to the final one and be able to tell if/how much things improved after training.

### CODE REVIEW

File: litgpt/finetune/adapter.py [hunk 1]
```diff
@@ -211,7 +211,7 @@ def fit(
         f" {model.max_seq_length} and context length is {model.config.block_size}"
     )
 
-    validate(fabric, model, val_dataloader, tokenizer, dataclasses.replace(eval, max_iters=2), data)  # sanity check
+    val_loss = validate(fabric, model, val_dataloader, tokenizer, dataclasses.replace(eval, max_iters=2), data)  # sanity check
```
Comment (carmocca) [inline litgpt/finetune/adapter.py:214]: Notice that this only does 2 iters of validation. Showing a value here could be very misleading.

I wonder if people get confused because we show `n/a`. Maybe reverting to only showing the val loss value when we have computed it once would help in avoid this confusion.

File: litgpt/finetune/adapter.py [hunk 2]
```diff
@@ -294,13 +291,20 @@ def fit(
                 save_hyperparameters(setup, checkpoint_file.parent)
                 save_prompt_style(data.prompt_style, checkpoint_file.parent)
 
+        if step_count == max_steps:
+            fabric.print("Final validation ...")
```
Comment (carmocca) [inline litgpt/finetune/adapter.py:295]: With the print below, this is redundant

File: litgpt/finetune/adapter.py [hunk 3]
```diff
@@ -294,13 +291,20 @@ def fit(
                 save_hyperparameters(setup, checkpoint_file.parent)
                 save_prompt_style(data.prompt_style, checkpoint_file.parent)
 
+        if step_count == max_steps:
```
Comment (carmocca) [inline litgpt/finetune/adapter.py:294]: You would want to also check if you already validated this step

File: litgpt/finetune/adapter.py [hunk 4]
```diff
@@ -294,13 +291,20 @@ def fit(
...
+            fabric.print("Final validation ...")
+            train_loss = validate(fabric, model, train_dataloader, tokenizer, eval, data, print_output=False)
```
Comment (carmocca) [inline litgpt/finetune/adapter.py:296]: This is slow. Why not rely on the running loss that is already printed?

File: litgpt/finetune/adapter.py [hunk 5]
```diff
@@ -294,13 +291,20 @@ def fit(
...
 
 # the adapter "kv cache" cannot be initialized under `inference_mode`
 @torch.no_grad()
 def validate(
-    fabric: L.Fabric, model: GPT, val_dataloader: DataLoader, tokenizer: Tokenizer, eval: EvalArgs, data: DataModule
+    fabric: L.Fabric, model: GPT, val_dataloader: DataLoader, tokenizer: Tokenizer, eval: EvalArgs, data: DataModule, print_output: bool = True
 ) -> torch.Tensor:
-    fabric.print("Validating ...")
+    if print_output:
+        fabric.print("Validating ...")
```
Comment (carmocca) [inline litgpt/finetune/adapter.py:307]: One SWE tip is to avoid this pattern. Instead add the print to the callers that have `print_output=True` and remove the argument.

Similarly, the instruction part could become a separate function. That is if we want to keep it because we have a lot of users confused by it

File: litgpt/finetune/adapter.py (hunk 1 above)
Comment (rasbt) [reply litgpt/finetune/adapter.py:214]: > Notice that this only does 2 iters of validation. Showing a value here could be very misleading.

That's fair. Would you be ok to run `validate` with the `eval.max_iters` here as done later? I think knowing the original, initial validation loss is extremely helpful because often the most improvement is made during the first steps. And this way you know what the performance of the original model was.

Also, the user confusion was not necessarily the n/a but the fact that the validation loss was never calculated because the finetuning finished in fewer steps than `eval.max_iters`. Right now, triggering validation requires a user to know the number of steps in advance and then set the `eval.max_iters` to a number smaller than that.

File: litgpt/finetune/adapter.py (hunk 4 above)
Comment (rasbt) [reply litgpt/finetune/adapter.py:296]: > This is slow. Why not rely on the running loss that is already printed?

I agree, this it is slow, but it's "only" exactly 2 times as slow as running validate as before. Sure, we could use the training loss from the previous iteration but then you have the same issue you pointed out for the validation loss:

> Notice that this only does 2 iters of validation. Showing a value here could be very misleading.

In practice it is useful to compare training and validation loss fairly to know the degree of overfitting in my opinion. We can maybe add a flag: `skip_validation` with default to `False` when we are running speed benchmarks. But for real uses cases, I think that knowing the training and validation losses is important.

File: litgpt/finetune/adapter.py (hunk 5 above)
Comment (rasbt) [reply litgpt/finetune/adapter.py:307]: The problem here is that these are two different validation prints. The first one is to signal that the model enters the validation (this may take a while if the validation set is very large). The second one is checking the prompt. If you move the `fabric.print("Validating ...")` then the print would occur when the validation has already happened.

```bash
@torch.no_grad()
def validate(
    fabric: L.Fabric, model: GPT, val_dataloader: DataLoader, tokenizer: Tokenizer, eval: EvalArgs, data: DataModule, print_output: bool = True
) -> torch.Tensor:
    if print_output:
        fabric.print("Validating ...")
    model.eval()
    losses = torch.zeros(min(len(val_dataloader), eval.max_iters))
    for k, batch in enumerate(val_dataloader):
        if k >= eval.max_iters:
            break
        input_ids, targets = batch["input_ids"], batch["labels"]
        logits = model(input_ids)
        losses[k] = chunked_cross_entropy(logits[..., :-1, :], targets[..., 1:], chunk_size=0)

    val_loss = losses.mean()

    if print_output:
        # produce an example:
        instruction = "Recommend a movie for me to watch during the weekend and explain the reason."
        fabric.print(instruction)
        prompt = data.prompt_style.apply(instruction)
        encoded = tokenizer.encode(prompt, device=fabric.device)
        with fabric.init_tensor():
            # do not set `max_seq_length=max_returned_token` because memory is not a concern here
            model.set_kv_cache(batch_size=1)
        output = generate(
            model, encoded, max_returned_tokens=len(encoded) + eval.max_new_tokens, temperature=0.8, eos_id=tokenizer.eos_id
        )
        model.clear_kv_cache()
        output = tokenizer.decode(output)
        fabric.print(output)

    model.train()
    return val_loss
```

File: litgpt/finetune/adapter.py (hunk 3 above)
Comment (rasbt) [reply litgpt/finetune/adapter.py]: this was a bit tricky ... the least convoluted way for this was to introduce a new variable because otherwise you have to do a complicated check since the model is not validated after each step.

File: litgpt/finetune/adapter.py (hunk 2 above)
Comment (rasbt) [reply litgpt/finetune/adapter.py]: yeah removed this

**Comment by rasbt** [comment]
To address the "it's slow" concern, I introduced a new `--skip_validation` option if a user is interested in not doing any evaluation, e.g., for benchmarking training performance, or for hparam tuning and deferring the validation to a later point in time when all the checkpoints were generated.

But I'd say in 95% of the cases you want to know the training loss and validation loss during and at the end of training to gauge overfitting and to know how well the model is doing in general. 

What do you think?

**Comment by rasbt** [comment]
I greatly simplified this @carmocca . Let me know if that's ok now. It should introduce any changes to the defaults.

File: litgpt/finetune/lora.py [hunk 6]
```diff
@@ -429,4 +433,4 @@ def validate_args(train: TrainArgs, eval: EvalArgs) -> None:
 if __name__ == "__main__":
     torch.set_float32_matmul_precision("high")
 
-    CLI(setup)
+    CLI(setup)
```
Comment (rasbt) [inline litgpt/finetune/lora.py:436]:
```suggestion
    CLI(setup)
 
```

File: litgpt/finetune/adapter.py [hunk 7]
```diff
@@ -15,7 +15,7 @@
 from torch.utils.data import DataLoader
 from torchmetrics import RunningMean
 
-from litgpt.adapter import GPT, Block, Config, adapter_filter, mark_only_adapter_as_trainable
+from litgpt.adapter_v2 import GPT, Block, Config, adapter_filter, mark_only_adapter_v2_as_trainable
```
[the remark this answers is not in the document]
Comment (rasbt) [reply litgpt/finetune/adapter.py]: ouch 😓

File: config_hub/finetune/falcon-7b/lora.yaml [hunk 8]
```diff
@@ -114,6 +114,9 @@ eval:
   # Number of iterations (type: int, default: 100)
   max_iters: 100
 
+  # Whether to evaluate on the validation set at the beginning of the training
+  initial_validation: false
```
Comment (awaelchli) [inline config_hub/finetune/falcon-7b/lora.yaml:118]: Is it really needed to add an argument for this? Why not just use the initial sanity check as the initial validation loss?

**Comment by rasbt** [comment]
> Is it really needed to add an argument for this? Why not just use the initial sanity check as the initial validation loss?

I originally had that but others found it was misleading because it's calculated only on 2 batches and is not reliable. I then replaced it with the full loss calculation but then this was too slow. So the new argument is a compromise so that I can enable it for me but it doesn't slow it down for everyone.
